fix `nix shell` and namespace nesting with diverted chroot store

moben opened the pull request:
# Motivation

Fixes two problems when using a chroot store (i.e. if `/nix` can't be created):
1. `nix shell` fails if it has to read a `propagated-user-env-packages`, as it does not take into account the actual location of the store outside the mount namespace.
2. No user namespaces can be created inside a namespace/chroot that was created by `nix`



# Context

`nix shell` failing to read the env packages is obviously just a bug and a trivial fix.

By using `pivot_root` instead of `chroot`, it is possible for child processes to create their own user namespaces. I've needed this in the past when using `nix shell` to run a script that itself tries to call `nix`. This technically changes behavior, as it's quite easy to "escape" `chroot`, but not `pivot_root`. Obviously it's not intended as a security barrier here but as a convenience feature, but it's still worth mentioning.

This enables running another `nix` process inside a `nix shell` as well as running other binaries that need to create a user namespace, such as sandboxing tools.

# Checklist for maintainers

<!-- Contributors: please leave this as is -->

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] documentation in the internal API docs
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or incompatible change: updated release notes

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

edolstra commented on src/nix/run.cc:
```diff
@@ -121,7 +124,7 @@ struct CmdShell : InstallablesCommand, MixEnvironment
             auto propPath = store->printStorePath(path) + "/nix-support/propagated-user-env-packages";
             if (accessor->stat(propPath).type == FSAccessor::tRegular) {
-                for (auto & p : tokenizeString<Paths>(readFile(propPath)))
+                for (auto & p : tokenizeString<Paths>(readFile(store->toRealPath(propPath))))
```
Since we already have an `accessor`, can't we do:

```suggestion
                for (auto & p : tokenizeString<Paths>(accessor->readFile(propPath)))
```

moben replied:
Good point, tested and applied.

edolstra commented on src/nix/run.cc:
```diff
@@ -13,6 +13,9 @@
 #include <sys/mount.h>
+#include <sys/syscall.h>
+#include <unistd.h>
+#define pivot_root(new_root, put_old) (syscall(SYS_pivot_root, new_root, put_old))
```
We already have definitions of `pivot_root` in `src/libstore/build/local-derivation-goal.cc` and `src/libstore/build/derivation-goal.cc` (which appears to be unused). Maybe we can move these into a header.

moben replied:
Sure, I can move it if required. Any suggestions where to move to? I don't know what the conventions are. `src/libutil/util.hh` seems like a bit of a catch-all? But not sure if that's really better for such one-off defines.

moben replied:
Dropped the macro instead. I think it just ends up being cruft too often. As you noted, 50% of the syscall macros in the entire code base are unused :grin:

I did make it consistent though, and there is already one place that calls `syscall(SYS_kill, ...)` for something platform specific on macos.

moben committed:
consistently use syscall directly where needed

There are some places in the code where a macro is defined and some where
`syscall` is used directly.

I think for these calls that don't have a lot of parameters and that are
only used in a few select places the macro is just cruft that gets left
behind as code moves around (see parent commit).

moben committed:
drop unused pivot_root macro

moben committed:
fix nested namespaces with diverted store

By using `pivot_root` instead of `chroot`, we serve the ability to
create nested user namespaces within our namespace.

This allows running sandboxing tools or just invoking `nix` again.

moben commented:
rebased to fix conflicts with master and address the last comment, though in a slightly different way. (one of the problems got fixed independently in the mean time)

thufschmitt commented:
Discussed during the Nix maintainers meeting on 2024-02-16.
Assigned to @thufschmitt.

<details>

Looks good, but a test will be welcome.
Assigned to @thufschmitt to help adding a test case
</details>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2024-02-16-nix-team-meeting-minutes-124/39870/1

arichtman-srt commented:
👀

edolstra commented:
Since Nix is using `pivot_root` now, the second issue addressed by this PR may not be current anymore? But the first one presumably still is. Could you have a look at updating it?

tomberek commented:
stale: re-open and re-base as needed
